### PR TITLE
Reduce allocations in blockmap code

### PR DESCRIPTION
**Pull request by calmh**
This doesn't actually do ~~much~~anything to improve the run time, but fewer allocations are generally good.

```
Before: BenchmarkBlockMapAdd    10   689798584 ns/op   662380159 B/op   406612 allocs/op
After:  BenchmarkBlockMapAdd    10   684422101 ns/op   105709137 B/op   126757 allocs/op
```

The reason I'm fiddling around in there is that a `fs.Update(protocol.LocalDeviceID, []protocol.FileInfo{someQuiteLargeFile})` is heavily limited by the blockmap code, taking almost 0.7 seconds in the benchmark for a 100.000 block file. The remaining 100k allocations are those done by `batch.Put()` which is hard to work around.

### CODE REVIEW

File: internal/db/blockmap_test.go [hunk 1]
```diff
@@ -20,11 +21,20 @@ func genBlocks(n int) []protocol.BlockInfo {
 	b := make([]protocol.BlockInfo, n)
 	for i := range b {
 		h := make([]byte, 32)
-		for j := range h {
-			h[j] = byte(i + j)
+		for i := 0; i < len(h)/4; i += 4 {
```
Comment (calmh) [inline internal/db/blockmap_test.go]: This is all to generate blocks with actual random hashes, instead of just a few hashes repeated.

**Comment by calmh** [comment]
Should probably do the same for Update() as well (so setting the `wait` label for now), and perhaps consider something to make this faster in general. Have any ideas?

**Comment by AudriusButkevicius** [comment]
So we can definatelt reduce allocations and do batching in other methods. I don't see how we could make it any waster though, as we have to work on a per-block level.

I guess we could also make some improvements where the blockmap is used, as now we sometimes drop the whole map and recreate it (somewhere in ldb code)

**Comment by calmh** [comment]
It needs benchmarking, I was thinking about thing like bucketing blocks by first 16 or so bits, so we get larger blobs of data to rewrite but limit it to 65k keys instead of millions etc.

**Comment by calmh** [comment]
i'll be baak

**Pull request closed without merging**